### PR TITLE
11.0 doc ged

### PR DESCRIPTION
**Pull request by mgeubelle**
Description of the issue/feature this PR addresses:

Current behavior before PR:

Desired behavior after PR is merged:

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

**Comment by Yenthe666** [comment]
Ooooh, the long awaited JavaScript documentation! 
Good job @ged-odoo 😉

**Comment by ged-odoo** [comment]
@Yenthe666 this is only the beginning.  When this is merged, we'll have a foundation, and every time an issue arises, we will update the documentation.  I hope that the documentation will be in a good state when we get to Odoo 12.0!

**Comment by Yenthe666** [comment]
@ged-odoo,

That is honestly great to hear. I believe documentation is underestimated at Odoo but it looks like you're doing a great effort to change this.
Thanks a lot for the great work 👏

File: doc/reference/javascript_reference.rst [hunk 1]
```diff
@@ -0,0 +1,1806 @@
...
+    removed sometimes in the future.
+
+.. attribute:: custom_events
+
+    this is almost the sams as the *events* attribute, but the keys
+    are arbitrary strings.  They represent business events triggered by
+    some sub widgets.  When an event is triggered, it will 'bubble up' the widget
+    tree (see the section on component communication for more details)
+
+.. function:: Widget.isDestroyed()
```
Comment (mge-odoo) [inline doc/reference/javascript_reference.rst]: what about `alive` method?

File: doc/reference/javascript_reference.rst [hunk 2]
```diff
@@ -0,0 +1,1806 @@
...
+  times in a single page (an example in Odoo is dashboards), queries should be
+  restricted to a given component's scope. Unfiltered selections such as
+  ``$(selector)`` or ``document.querySelectorAll(selector)`` will generally
+  lead to unintended or incorrect behavior.  Odoo Web's
+  :class:`~Widget` has an attribute providing its DOM root
+  (:attr:`~Widget.$el`), and a shortcut to select nodes directly
+  (:func:`~Widget.$`).
+
+* More generally, never assume your components own or controls anything beyond
+  its own personal :attr:`~Widget.$el`
```
Comment (mge-odoo) [inline doc/reference/javascript_reference.rst]: (and thus avoid using `this.parent` or `this.getParent()`?

File: doc/reference/javascript_reference.rst [hunk 3]
```diff
@@ -0,0 +1,1806 @@
...
+
+* html templating/rendering should use QWeb unless absolutely trivial.
```
Comment (mge-odoo) [inline doc/reference/javascript_reference.rst]: **H**tml

File: doc/reference/javascript_reference.rst [hunk 4]
```diff
@@ -0,0 +1,1806 @@
...
+* All interactive components (components displaying information to the screen
+  or intercepting DOM events) must inherit from :class:`~Widget`
+  and correctly implement and use its API and life cycle.
+
+
+QWeb Template Engine
+====================
+
+The web client uses the Qweb template engine to render widgets (unless they
+override the *renderElement* method to do something else). *LINK TO QWEB REF*
```
Comment (mge-odoo) [inline doc/reference/javascript_reference.rst]: don't forget the link

File: doc/reference/javascript_reference.rst [hunk 5]
```diff
@@ -0,0 +1,1806 @@
...
+With this, the *Counter* widget will load the xmlDependencies files in its
+*willStart* method, so the template will be ready when the rendering is performed.
+
+
+Event system
+============
+
+There are currently two event systems supported by Odoo: a simple system which
+allows adding listeners and triggering events, and a more complete system that
+also make events 'bubble up'.
```
Comment (mge-odoo) [inline doc/reference/javascript_reference.rst]: make**s**

File: doc/reference/javascript_reference.rst [hunk 6]
```diff
@@ -0,0 +1,1806 @@
...
+Here is an example on how this event system could be used:
+
+.. code-block:: javascript
+
+    var Widget = require('web.Widget');
+    var Counter = require('myModule.Counter');
+
+    var MyWidget = Widget.extend({
+        start: function () {
+            this.counter = new Counter(this);
```
Comment (mge-odoo) [inline doc/reference/javascript_reference.rst]: _super is not called here

File: doc/reference/javascript_reference.rst [hunk 7]
```diff
@@ -0,0 +1,1806 @@
...
+            // do something with val
+        },
+    });
+
+    // in Counter widget, we need to call the trigger method:
+
+    ... this.trigger('valuechange', someValue);
+
+
+Note: the use of this event system is discouraged, we plan to replace each
```
Comment (mge-odoo) [inline doc/reference/javascript_reference.rst]: maybe add this at the beginning of the section?

File: doc/reference/javascript_reference.rst [hunk 8]
```diff
@@ -0,0 +1,1806 @@
...
+*trigger* method by the *trigger_up* method from the extended event system
+
+Extended Event System
```
Comment (mge-odoo) [inline doc/reference/javascript_reference.rst]: Extended **e**vent **s**ystem

File: doc/reference/javascript_reference.rst [hunk 9]
```diff
@@ -0,0 +1,1806 @@
...
+---------------------
+
+The custom event widgets is a more advanced system, which mimic the DOM events
```
Comment (mge-odoo) [inline doc/reference/javascript_reference.rst]: The `custom_event` key in Widget ?

File: doc/reference/javascript_reference.rst [hunk 10]
```diff
@@ -0,0 +1,1806 @@
...
+API.  Whenever an event is triggered, it will 'bubble up' the component tree,
+until it reach the root widget, or is stopped.
```
Comment (mge-odoo) [inline doc/reference/javascript_reference.rst]: reach**es**

File: doc/reference/javascript_reference.rst [hunk 11]
```diff
@@ -0,0 +1,1806 @@
...
+
+    var Widget = require('web.Widget');
+    var Counter = require('myModule.Counter');
+
+    var MyWidget = Widget.extend({
+        custom_events: {
+            valuechange: '_onValueChange'
+        },
+        start: function () {
+            this.counter = new Counter(this);
```
Comment (mge-odoo) [inline doc/reference/javascript_reference.rst]: same remark

File: doc/reference/javascript_reference.rst [hunk 12]
```diff
@@ -0,0 +1,1806 @@
...
+            return this.counter.appendTo(this.$el);
+        },
+        _onValueChange: function(event) {
```
Comment (mge-odoo) [inline doc/reference/javascript_reference.rst]: same remark

File: doc/reference/javascript_reference.rst [hunk 13]
```diff
@@ -0,0 +1,1806 @@
...
+    Note that each value should be a subclass of *AbstractField*
+
+- view registry: this registry contains all JS views known to the web client
+    (and in particular, the view manager).  Each value of this registry should
+    be a subclass of *AbstractView*
+
+- action registry: we keep track of all client actions in this registry.  This
+    is where the action manager look up whenever it needs to create a client
+    action.  In version 11, each value should simply be a subclass of *Widget*.
+    However, in version 12, the values are required to be *AbstractAction*.
```
Comment (mge-odoo) [inline doc/reference/javascript_reference.rst]: As this is the documentation fort v11, I wouldn't mention this here, but I would do a commit in master to change it.

File: doc/reference/javascript_reference.rst [hunk 14]
```diff
@@ -0,0 +1,1806 @@
...
+        fieldRegistry.add('pad', FieldPad);
+
+    Note that each value should be a subclass of *AbstractField*
+
+- view registry: this registry contains all JS views known to the web client
+    (and in particular, the view manager).  Each value of this registry should
+    be a subclass of *AbstractView*
+
+- action registry: we keep track of all client actions in this registry.  This
+    is where the action manager look up whenever it needs to create a client
```
Comment (mge-odoo) [inline doc/reference/javascript_reference.rst]: look**s**

File: doc/reference/javascript_reference.rst [hunk 15]
```diff
@@ -0,0 +1,1806 @@
...
+                'open_record': '_onOpenRecord',
+            },
+            _onOpenRecord: function (event) {
+                var record = event.data.record;
+                var id = event.data.id;
+                // do something with the event.
+            },
+        });
+
+* Cross component
```
Comment (mge-odoo) [inline doc/reference/javascript_reference.rst]: component:

File: doc/reference/javascript_reference.rst [hunk 16]
```diff
@@ -0,0 +1,1806 @@
...
+        });
+
+        // in WidgetB
+        var WidgetB = Widget.extend({
+            ...
+            someFunction: function (barcode) {
+                core.bus.trigger('barcode_scanned', barcode);
+            },
+        });
+
```
Comment (mge-odoo) [inline doc/reference/javascript_reference.rst]: Maybe explain that new buses can be created if one doesn't want to use the one from core?

File: doc/reference/javascript_reference.rst [hunk 17]
```diff
@@ -0,0 +1,1806 @@
...
+--------
+
+A service is an instance of the *AbstractService* class.  It basically only has
+a name and a few methods.  Its job is to perform some work, typically something
+depending on the environment.
+
+For example, we have the *ajax* service (job is to perform a rpc), the
+*localStorage* (interact with the browser local storage) and many others.
+
+As an example, here is an simplified example on how the ajax service is implemented:
```
Comment (mge-odoo) [inline doc/reference/javascript_reference.rst]: Remove **As an example**

Comment (mge-odoo) [reply doc/reference/javascript_reference.rst]: Here is **a** simplified

File: doc/reference/javascript_reference.rst [hunk 18]
```diff
@@ -0,0 +1,1806 @@
...
+For services to work, it is necessary that we have a service provider ready to
+dispatch the custom events.  In the *backend* (web client), this is done by the
+main web client instance. Note that the code for the service provider comes from
+the *ServiceProviderMixin*.
+
+
+Widget
+------
+
+The widget is the part that request a service.  In order to do that, it simply
```
Comment (mge-odoo) [inline doc/reference/javascript_reference.rst]: request**s**

File: doc/reference/javascript_reference.rst [hunk 19]
```diff
@@ -0,0 +1,1806 @@
...
+trigger an event *call_service* (typically by using the helper function *call*).
```
Comment (mge-odoo) [inline doc/reference/javascript_reference.rst]: trigger**s**

File: doc/reference/javascript_reference.rst [hunk 20]
```diff
@@ -0,0 +1,1806 @@
...
+This event will bubble up and communicate the intent to the rest of the system.
+
+In practice, some functions are so frequently called that we have some helpers
```
Comment (mge-odoo) [inline doc/reference/javascript_reference.rst]: "helper functions" instead?

File: doc/reference/javascript_reference.rst [hunk 21]
```diff
@@ -0,0 +1,1806 @@
...
+function to make it easier to use. For example, the *_rpc* method is a helper
```
Comment (mge-odoo) [inline doc/reference/javascript_reference.rst]: make **them**

File: doc/reference/javascript_reference.rst [hunk 22]
```diff
@@ -0,0 +1,1806 @@
...
+    return this._rpc({
+        route: '/some/route/',
+        params: { some: kwargs},
+    });
+
+
+Translation management
+======================
+
+Some translations are made on the server side (basically all text rendered or
```
Comment (mge-odoo) [inline doc/reference/javascript_reference.rst]: text**s**

File: doc/reference/javascript_reference.rst [hunk 23]
```diff
@@ -0,0 +1,1806 @@
...
+  the JS module *web.core*
+- these strings are used by the server to generate the proper PO files
+- whenever the web client is loaded, it will call the route */web/webclient/translations*,
+  which returns a list of all translatable terms
+- in runtime, whenever the function *_t* is called, it will look up in this list
+  in order to find a translation, and return it or the original string if none
+  is found.
+
+Note that translations are explained in more details, from the server point of
+view, in [LINK].
```
Comment (mge-odoo) [inline doc/reference/javascript_reference.rst]: todo

File: doc/reference/javascript_reference.rst [hunk 24]
```diff
@@ -0,0 +1,1806 @@
...
+main need was to separate the rendering logic from the model logic.
+
+
+Views (in a generic sense) are now described with  4 pieces: a View, a
+Controller, a Renderer and a Model.  The API of these 4 pieces is described in
+the AbstractView, AbstractController, AbstractRenderer and AbstractModel classes.
+
+.. raw:: html
+
+    <svg width="550" height="173">
```
Comment (mge-odoo) [inline doc/reference/javascript_reference.rst]: what ?

File: doc/reference/javascript_reference.rst [hunk 25]
```diff
@@ -0,0 +1,1806 @@
...
+        <text xml:space="preserve" text-anchor="start" font-family="Helvetica, Arial, sans-serif" font-size="24" id="svg_11" y="44.5" x="346" fill-opacity="null" stroke-opacity="null" stroke-width="0" stroke="#000" fill="#000000">Controller</text>
+        <text xml:space="preserve" text-anchor="start" font-family="Helvetica, Arial, sans-serif" font-size="24" id="svg_12" y="128.5" x="276" fill-opacity="null" stroke-opacity="null" stroke-width="0" stroke="#000" fill="#000000">Renderer</text>
+        <text xml:space="preserve" text-anchor="start" font-family="Helvetica, Arial, sans-serif" font-size="24" id="svg_13" y="127.5" x="442" fill-opacity="null" stroke-opacity="null" stroke-width="0" stroke="#000" fill="#000000">Model</text>
+    </svg>
+
+- the View is the factory. Its job is to get a set of fields, arch, context and
+  some other parameters, then to construct a Controller/Renderer/Model triplet.
+
+  The view's role is to properly setup each other piece, with the correct
```
Comment (mge-odoo) [inline doc/reference/javascript_reference.rst]: piece**s**

File: doc/reference/javascript_reference.rst [hunk 26]
```diff
@@ -0,0 +1,1806 @@
...
+  information.  Usually, it has to process the arch string and extract the
+  data necessary for each other parts of the view.
+
+  Note that the view is a class, not a widget.  Once its job has been done, it
+  can be discarded.
+
+- the Renderer has one job: represent the data being viewed in a DOM element.
```
Comment (mge-odoo) [inline doc/reference/javascript_reference.rst]: represent**ing**

File: doc/reference/javascript_reference.rst [hunk 27]
```diff
@@ -0,0 +1,1806 @@
...
+
+  The Renderer is the V in the MVC pattern.
+
+- the Model: its job is to fetch and hold the state of the view.  Usually, it
+  represents in some way a set of records in the database.  The Model is the
+  owner of the 'business data'. It is the M in the MVC pattern.
+
+- the Controller: its job is to coordinate the renderer and the model.  Also, it
+  is the main entry point for the rest of the web client.  For example, when
+  the user change something in the search view, the *update* method of the
```
Comment (mge-odoo) [inline doc/reference/javascript_reference.rst]: change**s**

File: doc/reference/javascript_reference.rst [hunk 28]
```diff
@@ -0,0 +1,1806 @@
...
+  controller will be called with the appropriate information.
+
+  It is the C in the MVC pattern.
+
+Note: the JS code for the views has been designed to be usable outside of the
+context of a view manager/action manager.  They could be used in a client action,
+or, they could be displayed in the public website (with some work on the assets).
```
Comment (mge-odoo) [inline doc/reference/javascript_reference.rst]: Can we really state this in v11?

File: doc/reference/javascript_reference.rst [hunk 29]
```diff
@@ -0,0 +1,1806 @@
...
+
+Field Widgets
+=============
+
+A good part of the web client experience is about editing and creating data. Most
+of that work is done with the help of widgets, which are aware of the field
```
Comment (mge-odoo) [inline doc/reference/javascript_reference.rst]: the help of **field** widgets

File: doc/reference/javascript_reference.rst [hunk 30]
```diff
@@ -0,0 +1,1806 @@
...
+-------------
+
+The *AbstractField* class is the base class for all widgets in a view, for all
+views that support them (currently: Form, List, Kanban).
+
+There are many differences between the v11 field widgets and the previous versions.
+Let us mention the most important ones:
+
+- the widgets are shared between all views (well, Form/List/Kanban). No need to
+  duplicate the implementation anymore.
```
Comment (mge-odoo) [inline doc/reference/javascript_reference.rst]: maybe mention that if one wants a specific implementation in a view type, it can be done by prefixing the key by the view type in the field widgets registry

File: doc/reference/javascript_reference.rst [hunk 31]
```diff
@@ -0,0 +1,1806 @@
...
+    - Supported field types: *char*
+
+- date (FieldDate)
+    This is the default field type for fields of type *date*. Note that it also
+    works with datetime fields.  It uses the session timezone when formatting
+    dates.
+
+    - Supported field types: *date*, *datetime*
+
+- datetime (FieldDateTime0
```
Comment (mge-odoo) [inline doc/reference/javascript_reference.rst]: (FieldDateTime)

File: doc/reference/javascript_reference.rst [hunk 32]
```diff
@@ -0,0 +1,1806 @@
...
+
+- text (FieldText)
+    This is the default field type for fields of type *text*.
+
+    - Supported field types: *text*
+
+
+- handle (HandleWidget)
+    This field's job is to be displayed as a *handle* in a list view, and allow
+    reordering the various records to the desired state.
```
Comment (mge-odoo) [inline doc/reference/javascript_reference.rst]: to the desired state? Maybe rephrase this sentence

File: doc/reference/javascript_reference.rst [hunk 33]
```diff
@@ -0,0 +1,1806 @@
...
+- phone (FieldPhone)
+    This field displays a phone number.  The main reason to use it is that it
+    is rendered as an anchor tag with the proper href, in readonly mode, but
+    only in some cases: we only want to make it clickable if the device can
+    call this particular number.
+
+    - Supported field types: *char*
+
+- url (UrlWidget)
+    This field displays an url (in readonly mode). The url is represented as an  The main reason to use it is that it
```
Comment (mge-odoo) [inline doc/reference/javascript_reference.rst]: as an ... ?
Also go to next line before "The main reason"

File: doc/reference/javascript_reference.rst [hunk 34]
```diff
@@ -0,0 +1,1806 @@
...
+    .. code-block:: xml
+
+        <field name="active" widget="boolean_button" options='{"terminology": "archive"}'/>
+
+- boolean_toggle (BooleanToggle)
+    Displays a toggle switch to represent a boolean. This is a subfield of
+    FieldBoolean, mostly used to have a different look.
+
+- statinfo (StatInfo)
+    This widget is meant to represent statistical informations in a *stat button*.
```
Comment (mge-odoo) [inline doc/reference/javascript_reference.rst]: I think information has no plural form in English.

**Comment by mge-odoo** [comment]
The title "Javascript API" in the left menu seems wrongly aligned.
